snap/hooks/configure

```diff
@@ -1,158 +1,90 @@
-      echo "---"
-      exit 1
-    fi
-  else
-    if [ "${display_layout}" != "default" ]; then
-      echo "ERROR: '$display_layout' is not a layout in ${kiosk_display}:"
-      echo "---"
-      cat "${kiosk_display}"
-      echo "---"
-      exit 1
-    fi
-  fi
-
-  if [ "display-layout=${display_layout}" != "$(grep \^display-layout= ${kiosk_config})" ]; then
-    if [ $config_changes -gt 0 ]; then
-      sed '/^display-layout=/d' ${kiosk_config}
-    else
-      sed --in-place=.save '/^display-layout=/d' ${kiosk_config}
-    fi
-    echo display-layout=${display_layout} >> ${kiosk_config}
+# $3
+$1=${2}
+EOT
     let config_changes+=1
   fi
-fi
+}
 
-# vt
-vt=$(snapctl get vt)
-if [ -n "${vt}" ]; then
-  if [ -n "${vt//[0-9]}" ] || [ ! -e "/dev/tty$vt" ]; then
-    echo "ERROR: '$vt' is not a valid VT"
-    exit 1
-  fi
+# display_file
+if kiosk_display=$(snapctl get kiosk-display); then
+  # A display configuration can be supplied through the kiosk-display option
+  if [ -n "${kiosk_display}" ]; then
+    cat <<EOT >> "${display_file}~"
+# DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
+# USE 'snap set $SNAP_NAME kiosk-display=...' INSTEAD
 
-  if [ "vt=${vt}" != "$(grep vt= ${kiosk_config})" ]; then
-    if [ $config_changes -gt 0 ]; then
-      sed '/^vt/d' ${kiosk_config}
+${kiosk_display}
+EOT
+    if ! diff "${display_file}~" "${display_file}" > /dev/null; then
+      mv "${display_file}"  "${display_file}.save"
+      mv "${display_file}~" "${display_file}"
+      let config_changes+=1
     else
-      sed --in-place=.save '/^vt/d' ${kiosk_config}
+      rm "${display_file}~"
     fi
-    echo vt=${vt} >> ${kiosk_config}
-    let config_changes+=1
   fi
 fi
 
-# cursor
-cursor=$(snapctl get cursor)
-if [ -n "${cursor}" ]; then
-  if [[ ! "${cursor}" =~ ^(auto|none|software)$ ]]; then
-    echo "ERROR: '$cursor' is not a valid cursor option (auto|none|software)"
-    exit 1
-  fi
+# config_file
+if kiosk_config=$(snapctl get kiosk-config); then
+  # A kiosk configuration can be supplied through the kiosk-config option
+  if [ -n "${kiosk_config}" ]; then
+    cat <<EOT >> "${config_file}~"
+# DO NOT EDIT THIS FILE BY HAND -- YOUR CHANGES WILL BE OVERWRITTEN
+# USE 'snap set $SNAP_NAME kiosk-config=...' INSTEAD
 
-  if [ "${cursor}" == "none" ]; then
-    cursor="null"
-  fi
-
-  if [ "cursor=${cursor}" != "$(grep cursor= ${kiosk_config})" ]; then
-    if [ $config_changes -gt 0 ]; then
-      sed '/^cursor/d' ${kiosk_config}
+${kiosk_config}
+EOT
+    if ! diff "${config_file}~" "${config_file}" > /dev/null; then
+      mv "${config_file}"  "${config_file}.save"
+      mv "${config_file}~" "${config_file}"
+      let config_changes+=1
     else
-      sed --in-place=.save '/^cursor/d' ${kiosk_config}
+      rm "${config_file}~"
     fi
-    echo cursor=${cursor} >> ${kiosk_config}
-    let config_changes+=1
-  fi
-fi
+  else
+    # No kiosk configuration is supplied, we'll generate a default
```

Comment:
Should we `snapctl set kiosk-config` to the default, then?

Or is the idea that until the user `snap set`s, we'll always regenerate the defaults?

Reply:
The idea is that we merge the options we need with those in the config unless overridden by a snap option.

Reply:
This feels fragile. I'd rather the config file was owned by the configure hook.